extensions/amp-image-viewer/0.1/amp-image-viewer.js

```diff
@@ -261,104 +245,112 @@ export class AmpImageViewer extends AMP.BaseElement {
+    const img = elementByTag(ampImg, 'img');
+    this.measureElement(() => {
+      if (img) {
+        this.sourceWidth_ = img.naturalWidth || ampImg./*OK*/offsetWidth;
+        this.sourceHeight_ = img.naturalHeight || ampImg./*OK*/offsetHeight;
+      } else {
+        this.sourceWidth_ = ampImg./*OK*/offsetWidth;
+        this.sourceHeight_ = ampImg./*OK*/offsetHeight;
+      }
+    });
   }
 
   /**
    * Initializes the image viewer to the target image element such as
    * "amp-img". The target image element may or may not yet have the img
    * element initialized.
+   * @returns {!Promise}
    */
   init_() {
-    this.sourceWidth_ = this.getSourceWidth_();
-    this.sourceHeight_ = this.getSourceHeight_();
-    this.srcset_ = srcsetFromElement(dev().assertElement(this.sourceAmpImage_));
+    if (this.image_) {
+      return Promise.resolve();
+    }
 
-    ARIA_ATTRIBUTES.forEach(key => {
-      if (this.sourceAmpImage_.hasAttribute(key)) {
-        this.image_.setAttribute(key, this.sourceAmpImage_.getAttribute(key));
-      }
-    });
-    st.setStyles(dev().assertElement(this.image_), {
-      top: st.px(0),
-      left: st.px(0),
-      width: st.px(0),
-      height: st.px(0),
+    this.image_ = this.element.ownerDocument.createElement('img');
+    this.image_.classList.add('i-amphtml-image-viewer-image');
+    const ampImg = dev().assertElement(this.sourceAmpImage_);
+    this.setSourceDimensions_(ampImg);
+    this.srcset_ = srcsetFromElement(ampImg);
+
+    return this.mutateElement(() => {
+      st.setStyles(dev().assertElement(this.image_), {
+        top: 0,
+        left: 0,
+        width: 0,
+        height: 0,
+      });
+      st.toggle(ampImg, false);
+      this.element.appendChild(this.image_);
+      return ampImg.getImpl().then(ampImg => {
+        ampImg.propagateAttributes(ARIA_ATTRIBUTES, this.image_);
+      });
     });
   }
 
   /**
    * Measures the image viewer and image sizes and positioning.
    * This must be called AFTER the source element has already been
    * laid out.
+   * @private
+   */
+  measure_() {
+    this.elementBox_ = layoutRectFromDomRect(this.element
+        ./*OK*/getBoundingClientRect());
+
+    const sourceAspectRatio = this.sourceWidth_ / this.sourceHeight_;
+    let height = Math.min(this.elementBox_.width / sourceAspectRatio,
+        this.elementBox_.height);
+    let width = Math.min(this.elementBox_.height * sourceAspectRatio,
+        this.elementBox_.width);
+
+    if (Math.abs(width - this.sourceWidth_) <= 16
+    && Math.abs(height - this.sourceHeight_ <= 16)) {
+      width = this.sourceWidth_;
+      height = this.sourceHeight_;
+    }
+
+    this.imageBox_ = layoutRectLtwh(
+        Math.round((this.elementBox_.width - width) / 2),
+        Math.round((this.elementBox_.height - height) / 2),
+        Math.round(width),
+        Math.round(height));
+
+    // Adjust max scale to at least fit the screen.
+    const elementBoxRatio = this.elementBox_.width
```

Comment:
nit: the line break should come after the `/`. there's a few places in this PR that break before the operator

https://google.github.io/styleguide/jsguide.html#formatting-where-to-break

> Operators are wrapped as follows:
> When a line is broken at an operator the break comes after the symbol